### PR TITLE
Add Keplearn: deterministic symbolic regression by best-first recursive reduction

### PR DESCRIPTION
**Pull request by owls-on-wires**
This PR adds Keplearn, a symbolic regression engine written in Rust
(https://github.com/owls-on-wires/keplearn, MIT). It recovers closed-form
equations by reducing the target through chains of invertible operations
(divide out a fitted factor, invert a transform, fit an inner affine node)
until a terminal monomial or composite fit explains the reduced target.
Candidates are selected on an internal holdout with an MDL complexity
penalty. The engine contains no randomness; runs are reproducible byte for
byte.

### Self-reported results

All numbers below come from this repository's containerized pipeline
(`evaluate_model.py` splits and seeds, `assess_symbolic_model` grading) at
one trial per dataset and a 60 second per-fit cap. That budget is well
under the official 3600 s / 36000 s settings, so these are a lower bound
rather than official results; we state it up front to be clear about
compute.

Ground truth, symbolic solution rate by target noise:

| suite | noise 0 | 0.001 | 0.01 | 0.1 |
|---|---|---|---|---|
| Feynman (116) | 75 (64.7%) | 65 (56.0%) | 51 (44.0%) | 10 (8.6%) |
| Strogatz (14) | 6 (42.9%) | 6 (42.9%) | 3 (21.4%) | 1 (7.1%) |

Ground truth, accuracy solutions (R2_test >= 0.999) by target noise:

| suite | noise 0 | 0.001 | 0.01 | 0.1 |
|---|---|---|---|---|
| Feynman (116) | 92 (79.3%) | 86 (74.1%) | 74 (63.8%) | 15 (12.9%) |
| Strogatz (14) | 7 (50.0%) | 7 (50.0%) | 5 (35.7%) | 1 (7.1%) |

Black-box (138 datasets, no ground-truth form; accuracy only):

| metric | value |
|---|---|
| median R2_test | -0.003 |
| mean R2_test | -0.34 |
| positive R2_test (> 0) | 32/138 (23.2%) |

Fit time under the 60 s cap (one core per fit):

| suite | median | mean | max | fits under 1 s |
|---|---|---|---|---|
| Feynman, noise 0 | 1.3 s | 13.3 s | 60.7 s | 55/116 |
| Feynman, noise 0.001-0.1 | 11-14 s | 15-23 s | 60.7 s | 0/348 |
| Strogatz, all noise levels | 1.3 s | 1.5 s | 3.6 s | 8/56 |
| Black-box | 60.0 s | 39.1 s | 65.8 s | 15/138 |

Solved instances exit early; at noise 0, half the Feynman suite finishes in
under a second. Under noise the exact-exit path cannot fire, so fits run to
frontier exhaustion instead (hence the higher medians there). The black-box
median sits at the cap, so those fits are budget-limited.

### Layout

- `algorithms/keplearn/`: `metadata.yml`, `environment.yml`, `install.sh`,
  `LICENSE`. `install.sh` clones the pinned release tag `v0.1.0` of the
  public repository and builds it with cargo from conda-forge; the build
  toolchain is removed again in the same layer, so the image stays near the
  base size. No source or binaries are vendored in this PR.
- `experiment/methods/keplearn/`: `regressor.py`, `__init__.py`. The
  estimator shells out to the installed binary (training data as TSV on
  stdin, model as JSON on stdout). `model(est, X)` returns a
  sympy-compatible string in the columns of `X`.

### Contract notes

- `est.max_time` is read at fit time and passed to the engine, which stops
  on the budget and emits its best model so far.
- `est.random_state` is accepted for API compatibility and has no effect:
  the engine is deterministic, including its internal holdout split. Across
  trials, variation comes only from the harness's data splits and noise
  seeds.
- `eval_kwargs` sets `scale_x=False, scale_y=False, use_dataframe=False`;
  the engine fits raw features and bakes constants into the model string.
- Datasets with fewer than 40 rows return the constant model "0" (the
  internal holdout gate abstains rather than guess).

### Testing

- `python scripts/check_method_layout.py`: method layout OK.
- `docker compose build base && docker compose build keplearn`: the image
  self-installs from the public tag during the build.
- `docker compose run --rm keplearn bash test.sh`: `test_algorithm` passes
  3/3 (import, evaluate, sympy) and `test_evaluate_model` passes.

### Author

Chandler Freeman, contact at chandler@mnty.sh

### CODE REVIEW

**Comment by gAldeia** [comment]
Hi @owls-on-wires,

Thanks for creating this PR and for sharing the results you got on your end. Also, pretty impressive fit speed, and thanks for implementing support for `max_time` — this will be really useful in future experiments.

Personal comment: I like the idea of a deterministic search and using MDL as a penalty.

I also appreciate how you made SymPy part of your method's `predict()` and `model()` calls — it feels much more likely that we'll see compatible results with our current experimental pipeline.

There are a few issues that I believe need to be fixed before we can proceed with merging this PR.

**1. Clip and NaN handling.**

I think this is an interesting feature, but I'm not sure how it fits into our experimental pipeline, and I'd like some outside input (@lacava) to help make a decision here.

```python
# Predict-side hygiene: non-finite -> training mean; clip to the training target
# range so extrapolation blow-ups the selector cannot see can't wreck the score.
v = np.where(np.isfinite(v), v, self.y_mean_)
return np.clip(v, self.y_lo_, self.y_hi_)
```

**2. Model string.**

There's a simplification step when generating your model string and counting the model size. This can produce a model size that mismatches the actual model being used for prediction.

Also, since your prediction function has a protection wrapper to avoid out-of-domain predictions, that logic should be reflected in the model string, because we measure expression size based on the SymPy-compatible string. And since the expression is simplified in this context but not at prediction time, there's a chance the formula could get rewritten differently. We expect that `predict(X)` will always match evaluating `model(est, X)`.

**3. Model size.**

Your custom complexity function is used as a fallback when the experiment fails to generate a valid SymPy string. Although this is unlikely in your implementation (since SymPy is already part of your model), it would be nice for your custom complexity function to be as faithful as possible to the actual model used for prediction. That is also related with the potential mismatch generated by your simplification step.

I am again unsure about the clipping mechanism. My instinct is to also count those extra nodes used for clipping in the expression size, but I would love to have someone else's input on this

**Comment by owls-on-wires** [comment]
Hi @gAldeia, thank you for the review and kind comments! In the most recent update:

**Issues 2 and 3 are addressed.** The regressor now builds one canonical SymPy expression at fit time, and every consumer uses that same object: `predict` evaluates it (via `lambdify`), `model()` prints it with the feature-name substitution, and expression size is measured on it. The normalization pass now feeds predictions rather than only the reported string, so the mismatch you identified cannot occur anymore. I verified inside the container that `predict(X)` equals evaluating `model(est, X)` elementwise, at zero tolerance. The `complexity()` fallback also no longer returns 0; it now does a token-level operator count of the raw engine string, though as you noted that path should be unreachable in practice.

**On the clip: I tried removing it altogether.** `predict` returns the evaluated expression as-is wherever it is finite. Re-running the full suite through this repository's pipeline (one trial, 60 s per fit) with the clip gone:

Ground truth is unchanged on symbolic solutions and slightly better on accuracy, since predictions now come from the normalized expression:

| suite | metric | noise 0 | 0.001 | 0.01 | 0.1 |
|---|---|---|---|---|---|
| Feynman (116) | symbolic | 75 | 65 | 51 | 10 |
| Feynman (116) | accuracy | 93 (+1) | 87 (+1) | 75 (+1) | 16 (+1) |
| Strogatz (14) | symbolic | 6 | 6 | 3 | 1 |
| Strogatz (14) | accuracy | 7 | 7 | 5 | 1 |

Black-box is where it shows, with a mean R2_test of -4518:

| metric | with clip (as submitted) | clip removed |
|---|---|---|
| median R2_test | -0.003 | -0.004 |
| positive R2_test | 32/138 | 28/138 |
| mean R2_test | -0.34 | -4518 |

The median is unaffected and all 138 datasets still complete. The clip's effect was the tail: six datasets now score below -100, and one (634_fri_c2_100_10) reaches about -6e5. Those are the real numbers for models that extrapolate badly, so it is honest and I'm ok with that as a result (either way, black-box performance is not our strong suite).

If you'd like to take another direction, I'd be happy to change it if you and @lacava prefer.

Thank you!

**Comment by gAldeia** [comment]
Hi @owls-on-wires,

Thank you for reading my message and quickly addressing the issues.

I am glad that these changes did not substantially affected your results -- with the exception of the mean R2. Given that the R2 is unbounded negatively, the mean value can be pretty low (this was a common issue for many SR methods last time I ran the benchmark). On post-hoc analysis notebooks we usually clip them to a negative value to deal with this. Glad that the median didn't changed significantly.

For future iterations of the benchmark, we will consider some sort of mechanism to deal with potential invalid evaluations and use it across models if pertinent.

Happy to merge!
